prodsys/simulation/control.py

```diff
@@ -286,8 +286,190 @@ def run_process(self, input_state: state.State, target_product: product.Product)
             state.StateTypeEnum.production,
         )
         input_state.process = self.env.process(input_state.process_state())
+        # TODO: if performed process has failure_rate -> sample (np.choice) if failure happened, if so: mark product as rework_needed
         yield input_state.process
 
+class BatchController(Controller):
+    """
+    A production controller is responsible for controlling the processes of a production resource. The controller is requested by products requiring processes. The controller decides has a control policy that determines with which sequence requests are processed.
+    """
+    resource: resources.ProductionResource = Field(init=False, default=None)
+
+    def get_batch_size(self, resource: resources.Resource) -> int:
+        """
+        Get the batch size for the given resource.
+
+        Args:
+            resource (resources.Resource): The resource to get the batch size for.
+
+        Returns:
+            int: The batch size of the resource.
+        """
+        if isinstance(resource, resources.ProductionResource):
+            return resource.batch_size
+        else:
+            raise ValueError("Resource is not a ProductionResource")
+
+    def get_next_product_for_process(
+        self, resource: resources.Resource, process_request: request.Request
+    ) -> Generator:
+        """
+       Get the next batch of products for a process. The products are removed (get) from the input queues of the resource.
+
+        Args:
+            resource (resources.Resource): The resource to take the products from.
+            process_request (request.Request): The request that is requesting the products.
+
+        Returns:
+            List[events.Event]: The events that are triggered when the products are taken from the queue.
+        """
+
+        events = []
+        products = []
+        batch_size = self.get_batch_size(resource)
+        if isinstance(resource, resources.ProductionResource):
+            for queue in resource.input_queues:
+                while len(products) < batch_size:
+                    event = queue.get(filter=lambda item: item is process_request.get_product().product_data)
```

Comment:
I think this could be problematic, since the ID of product_data is changed for individual instances... so only one instance exist and other products are not selected. I think this could be a problem that batches contain the same product instance multiple times and not x items of product of one product type for x being the batch size

Reply:
Would this be correct instead to get the correct products?
event = queue.get(filter=lambda item: item.ID == process_request.get_product().product_data.ID)

Reply:
- Products should be a set of product.product_data_id
- The filter to the queue should check if the product_data.product_type equals that of the request and not item.product_data.ID is not in products set
- item.product_data.ID is added to the products set